### ISSUE TITLE
[Feature Request] Player.ChangeCharacterName(string characterName)

### ISSUE THREAD

**Issue by IAmSilK** [SmartlyDressedGames/Unturned-3.x-Community#2902]
Back again with another plugin developer request.

Something that has been sought after for a long time by plugin developers is the ability to change a player's name after they've already connected to the server. This is difficult to do at the moment as the only possible way is by replicating to clients a fake disconnect and reconnect a specific player in order to pass a new character name. This comes with all sorts of bugs as obviously this is unintended and a lot of work for such a simple feature.

Support in the form of some RPC which would replicate player character name changes after the player has connected to the server would be great.

Some bugs may appear where player character names are copied when the player connects and therefore won't replicate in some components, but as this would just be visual issues, these bugs would be non-fatal and can be dealt with as time goes on.

I'm sure many developers would appreciate this functionality being added to the game.

**Comment by SDGNelson** [comment]
Why do you want to change the player name?

**Comment by surv0013** [comment]
> Why do you want to change the player name?

He just wants... I want too.

**Comment by muertoof** [comment]
Yes, it is necessary for many developers. At the moment, we can create a lot of dynamic systems for different types of servers, but, for example, to update a nickname when assigning a new role or updating a rank, you still need to reconnecting to the server.

**Comment by IAmSilK** [comment]
> Why do you want to change the player name?
- Permission role prefixes/suffixes ([Owner], [Admin], etc.)
- RP job titles ([Farmer], [Cop], etc.)
- PVP rankings ([#1] Nelson) (these especially as they possibly change on every kill the fly)

Other devs will have their other reasons and few have already commented here.

**Comment by DanielWillett** [comment]
This would be a nice feature, also `ChangeNickName` as well... As Silk stated, many servers put ranks or permission prefixes in front of players' names, and, especially when there's a queue, players don't want to rejoin to update their names, since that's really the only time you can change names (in `Provider.onCheckValidWithExplanation`). Another good use is giving a player some kind of status in front of their name... for example: `[AFK] Mr. Smith` or `[HURT] Mr. Smith`.

**Comment by BarehSolok** [comment]
yessss

**Comment by fahrettinenes** [comment]
Sounds good. It can be so good for RolePlay server plugins.

**Comment by connorgillespie** [comment]
Agreed. This would be very useful for showing prefixes and rankings in front of players' names.

**Comment by Johnanater** [comment]
This would be an awesome feature.

**Comment by Patryk-0** [comment]
Agree,I don't know why this feature is not added yet,it will be very useful for developers when creating plugins

**Comment by zanarck78** [comment]
Hello @SDGNelson !!

Personaly l wanted this option a long time ago but never tryied to request this feature,for what we want these functions you said? well there is a lot of aplication for the nickname changing like prefixes for rp servers/pvp,real time pvp statistics and more...

Thanks for listening us! ;)

**Comment by IAmSilK** [comment]
@SDGNelson could we get an update on your thoughts?

**Comment by SDGNelson** [comment]
It might happen, but I will need to check how many places assume the character name does not change. e.g. UI that needs an event for name change.

**Comment by GazziFX** [comment]
Also there should be `OriginalName` property to easy change name without stacking prefixes

**Comment by IAmSilK** [comment]
> but I will need to check how many places assume the character name does not change
To help streamline the possibility of this feature being added, I've gone ahead and checked everywhere the character name is accessed and then persisted in a separate variable.

Important areas where name will need to be updated:
- `PlayerDashboardInformationUI.updateMarkers`
- `PlayerDashboardInformationUI.updateRemotePlayerAvatars`
- `PlayerDashboardInventoryUI.updateVehicle`
- `PlayerLifeUI.onVotingStart`
- `SteamGameServer.BUpdateUserData`
- `PlayerUI.updateGroupLabels`
- SleekPlayer's constructor (used in SleekPlayer.nameLabel)

Less important areas where its questionable if it needs to be updated:
- `PlayerDeathUI.synchronizeDeathCause`
- `PlayerLifeUI.onArenaPlayerUpdated`
- `PlayerNPCDialogueUI.updateDialogue`
- `PlayerNPCDialogueUI.updatePage`

**Comment by Patryk-0** [comment]
it's been almost a month from the publication of this post,any updates @SDGNelson ?

**Comment by IAmSilK** [comment]
Also the most upvoted issues post here 👍

**Comment by connorgillespie** [comment]
> it's been almost a month from the publication of this post,any updates @SDGNelson ?

Bump. Been excitedly waiting for this in the patch notes.

**Comment by surv0013** [comment]
We are waiting...

**Comment by GazziFX** [comment]
> We are waiting...

Im waiting sendItemImage

**Comment by IAmSilK** [comment]
Unrelated issue, but take a look here: https://github.com/SilKsPlugins/UnturnedImages

**Comment by GazziFX** [comment]
> Unrelated issue, but take a look here: https://github.com/SilKsPlugins/UnturnedImages

https://github.com/GazziFX/Unturned-Icons

**Comment by SDGNelson** [comment]
Closing because the issues list was not really intended for feature requests. Sorry to close this discussion - I do WANT to fulfill all (most?) requests, but it is a matter of prioritization. Constantly being reminded of all the pending requests when checking issues is stressful. I will keep the request in mind. On GitHub maybe the Ideas section of Discussions would be the best place.
